src/theory/evaluator.cpp

```diff
@@ -140,30 +158,103 @@ EvalResult Evaluator::evalInternal(TNode n,
+      {
+        // we cannot evaluate since there was an invalid child
         doEval = false;
       }
     }
+    Trace("evaluator") << "Evaluator: visit " << currNode
+                       << ", process = " << doProcess
+                       << ", evaluate = " << doEval << std::endl;
 
-    if (doEval)
+    if (doProcess)
     {
       queue.pop_back();
 
       Node currNodeVal = currNode;
+
+      // The code below should either:
+      // (1) store a valid EvalResult into results[currNode], or
+      // (2) store an invalid EvalResult into results[currNode] and
+      // store the result of substitution + rewriting currNode { args -> vals }
+      // into evalAsNode[currNode].
+
+      // If we did not successfully evaluate all children
+      if (!doEval)
+      {
+        // Reconstruct the node with a combination of the children that
+        // successfully evaluated, and the children that did not.
+        Trace("evaluator") << "Evaluator: collect arguments" << std::endl;
+        std::vector<Node> echildren;
+        if (currNode.getMetaKind() == kind::metakind::PARAMETERIZED)
+        {
+          echildren.push_back(currNode.getOperator());
+        }
+        for (const auto& currNodeChild : currNode)
+        {
+          itr = results.find(currNodeChild);
+          if (itr->second.d_tag == EvalResult::INVALID)
+          {
+            // could not evaluate this child, look in the node cache
+            itn = evalAsNode.find(currNodeChild);
+            if (itn != evalAsNode.end())
+            {
+              // take cached value
+              echildren.push_back(itn->second);
+            }
+            else
+            {
+              // must take original
+              echildren.push_back(currNodeChild);
+            }
+          }
+          else
+          {
+            // otherwise, use the evaluation
+            echildren.push_back(itr->second.toNode());
+          }
+        }
+        // The value is the result of our (partially) successful evaluation
+        // of the children.
+        currNodeVal = nm->mkNode(currNode.getKind(), echildren);
+        Trace("evaluator") << "Evaluator: partially evaluated " << currNodeVal
+                           << std::endl;
+        // Use rewriting. Notice we do not need to substitute here since
+        // all substitutions should already have been applied recursively.
+        currNodeVal = Rewriter::rewrite(currNodeVal);
+        Trace("evaluator") << "Evaluator: now after substitution + rewriting: "
+                           << currNodeVal << std::endl;
+        if (currNodeVal.getNumChildren() > 0)
+        {
+          // We may continue with a valid EvalResult at this point only if
+          // we have no children. We must otherwise fail here since some of
+          // our children may not have successful evaluations.
+          results[currNode] = EvalResult();
+          evalAsNode[currNode] = currNodeVal;
+          continue;
+        }
```

Comment:
Could you add a comment that we try to turn a `currNodeVal` with no children into an `EvalResult`? I was a bit confused by why we continue at first.